lib/serve-ts.js

```diff
@@ -0,0 +1,126 @@
+// @ts-check
+/* eslint-env node */
+
+const child_process = require('child_process');
+const fs = require('fs');
+
+const rimraf = require('rimraf');
+
+module.exports = (project, outDir) => {
+  const Serve = project.require('ember-cli/lib/commands/serve');
+  const Builder = project.require('ember-cli/lib/models/builder');
+  const Watcher = project.require('ember-cli/lib/models/watcher');
+
+  /**
+   * Exclude .ts files from being watched
+   */
+  function filterTS(name) {
+    if (name.startsWith('.')) {
+      // these files are filtered by default
+      return false;
+    }
+
+    // typescript sources are watched by `tsc --watch` instead
+    return !name.endsWith('.ts');
+  }
+
+  class WatcherNonTS extends Watcher {
+    buildOptions() {
+      let options = super.buildOptions();
+      options.filter = filterTS;
+      return options;
+    }
+  }
+
+  return Serve.extend({
+    name: 'serve-ts',
+    aliases: ['st'],
+    works: 'insideProject',
+    description:
+      'Serve the app/addon with the TypeScript compiler in incremental mode. (Much faster!)',
+
+    run(options) {
+      const config = this.project.config(options.environment);
+
+      this.project._isRunningServeTS = true;
+
+      return new Promise(resolve => {
+        let started = false;
+
+        this.ui.startProgress('Starting TypeScript compilation...');
+
+        // TODO: typescript might be installed globally?
+        // argument sequence here is meaningful; don't apply prettier.
+        // prettier-ignore
+        this.tsc = child_process.fork(
+          'node_modules/typescript/bin/tsc',
+          [
+            '--watch',
```

Comment:
@dwickern @dfreeman one thing I've been thinking is: we *should* be able to get and use the project's tsconfig, and then use that here (using `-p <path to project>`) except with overrides. I *believe* TS properly does the override sequence if you just hand it a project file and then a set of overrides.

Does that make sense to both of you?

Reply:
Alternative approach: do the usual `JSON.parse(tsConfig)` dance and set only the properties we want to set here, e.g. `sourceMap: <whatever is in the config` That might be better.

Also, I don't think we need to worry about using a global install of `tsc`; we should *only* be using the local version, since that's what will be used for the final build, as well.

Reply:
I like the idea of using `-p <path-to-project>` and overriding with the handful of flags we need — it feels simpler (and I also _think_ `tsc` will actually let you get away with nonconformant JSON, so we'd need something fancier than `JSON.parse` for that approach)

Reply:
I think it already picks up the tsconfig in the project root based on the cwd? We could pass the path explicitly though.

Reply:
Yeah, I just went back and read the docs, and you're both correct about those pieces. So here, we should just be able to override *only what we need to* – and presumably, other than outdir, that should largely match what we have in the overrides in the `TypescriptPreprocessor` class (starting [here](https://github.com/typed-ember/ember-cli-typescript/pull/85/files#diff-fa6ce9add76b6ec544e7e53030f16ddcR24)).

Note that *that's* assuming valid JSON. 🤔